docs/design/assignment.md

```diff
@@ -0,0 +1,329 @@
+
+```
+var n: i32;
+// Error, assignment is not permitted as a subexpression.
+if (F() and (n = GetValue()) > 5) {
+}
+```
+
+User-defined types can define the meaning of these operations by
+[implementing an interface](#extensibility) provided as part of the Carbon
+standard library.
+
+## Syntax
+
+The operands of these operators can be any [expression](expressions/README.md).
+However, the first operand must be modifiable because it is passed to an
+`[addr self: Self*]` parameter, which disallows most expression forms other
+than:
+
+-   The name of a `var` binding.
+-   A dereference of a pointer.
+-   Array indexing that produces a modifiable result.
+-   Member access naming a field, where the object is one of these expressions.
+-   A conditional expression selecting between these expressions.
+
+## Simple assignment semantics
+
+A simple assignment statement is intended to exactly mirror the semantics of
+initialization. The following two code snippets should have the same meaning if
+they are both valid:
+
+```
+// Declare and initialize.
+var v: T = init;
+```
+
+```
+// Declare separately from initialization.
+// Requires that `T` has an unformed state.
+var v: T;
+v = init;
+```
+
+This equivalence is not enforced, but when an object is in an unformed state,
+running the assignment function is _optional_, just like running the destructor
+is. If the assignment function is not run, the object will be directly
+initialized from the right-hand side instead. The type is still required to
+implement `AssignFrom` for the assignment to be valid.
+
+```
+class C { ... }
+fn F() -> C {
+  returned var c: C = {...};
+  // `&c` here is `&x` for the first call to `F()`.
+  // `&c` here can be `&y` for the second call  to `F()`.
+  return var;
+}
+fn G() {
+  var x: C = F();
+  var y: C;
+  y = F();
+}
+```
+
+## Built-in types
+
+For built-in types:
+
+-   For `=`, the right-hand operand is implicitly converted to the type of the
+    left-hand operand, and the converted value replaces the value of that
+    operand.
+-   A `$=` operator is supported in each case where a built-in `$` operator is
+    supported, and `a $= b;` behaves the same as `a = a $ b;` except that `a` is
+    only evaluated once.
+-   For integer types, `++n;` and `--n;` behave the same as `n += 1;` and
+    `n -= 1;` respectively. For floating-point types, these operators are not
+    provided.
+
+## Tuples, structs, and classes
+
+_TODO_: Describe the rules for assignment in these cases.
+
+See leads issue
+[#686: Operation order in struct/class assignment/initialization](https://github.com/carbon-language/carbon-lang/issues/686)
+
+## Extensibility
+
+Assignment operators can be provided for user-defined types by implementing the
+following families of interfaces. Implementations of these interfaces are
+provided for built-in types as necessary to give the semantics described above.
+
+### Simple assignment
+
+```
+// Simple `=`.
+interface AssignFrom(U:! type) {
```

Comment:
I find the difference of `AssignFrom` vs `AssignWith` a bit odd .. is there a wording that would work for both?

Reply:
Not fully satisfying, but maybe `AssignGiven` and `AddAssignGiven`

Reply:
After considering a bunch of options, I've settled on `Assign` / `AssignWith` / `AssignOp` / `AssignOpWith` being the most consistent set of names. I'm not entirely happy with these names, but I'm also not sure we can do better.

Reply:
After much discussion, reversed `AssignOp` back to `OpAssign`.